Scalar.Common/Git/GitCredentialManagerProcess.cs

```diff
@@ -0,0 +1,276 @@
+            {
+                return false;
+            }
+            stdin.Append("\n");
+
+            using (ITracer activity = tracer.StartActivity(nameof(this.TryGetCredential), EventLevel.Informational))
+            {
+                Result result = this.InvokeGcm("get", stdin.ToString());
+
+                if (!result.IsSuccess)
+                {
+                    tracer.RelatedError("GCM could not get credentials: {0}", result.Errors);
+                    errorMessage = result.Errors;
+                    return false;
+                }
+
+                username = ParseValue(result.Output, "username=");
+                password = ParseValue(result.Output, "password=");
+
+                bool success = username != null && password != null;
+
+                var metadata = new EventMetadata();
+                metadata.Add("Success", success);
+                if (!success)
+                {
+                    metadata.Add("Output", result.Output);
+                }
+
+                activity.Stop(metadata);
+                return success;
+            }
+        }
+
+        private static bool TryAppendUrlLines(StringBuilder sb, string url, out string errorMessage)
+        {
+            if (Uri.TryCreate(url, UriKind.Absolute, out Uri uri))
+            {
+                sb.AppendFormat("protocol={0}\n", uri.Scheme);
+                sb.AppendFormat("host={0}\n", uri.Host);
+
+                if (uri.PathAndQuery != null)
+                {
+                    string[] pathAndQuery = uri.PathAndQuery.Split('?');
+                    if (pathAndQuery.Length > 0)
+                    {
+                        // Trim any trailing "/" from the path
+                        string path = pathAndQuery[0]?.TrimEnd('/');
+                        if (!string.IsNullOrWhiteSpace(path))
+                        {
+                            sb.AppendFormat("path={0}\n", uri.PathAndQuery);
+                        }
+                    }
+                }
+
+                errorMessage = null;
+                return true;
+            }
+
+            errorMessage = $"Failed to parse URL '{url}' as a valid URI";
+            return false;
+        }
+
+        private static string ParseValue(string contents, string prefix)
+        {
+            int startIndex = contents.IndexOf(prefix, StringComparison.Ordinal) + prefix.Length;
+            if (startIndex >= 0 && startIndex < contents.Length)
+            {
+                int endIndex = contents.IndexOf('\n', startIndex);
+                if (endIndex >= 0 && endIndex < contents.Length)
+                {
+                    return
+                        contents
+                            .Substring(startIndex, endIndex - startIndex)
+                            .Trim('\r');
+                }
+            }
+
+            return null;
+        }
+
+        private Result InvokeGcm(string command, string stdIn)
```

Comment:
If this becomes `protected virtual`, then we could extend the class in the test suite and unit-test the expected value and frequency of calls to this method. Could be a helpful thing to do, considering we shouldn't ever be in this mode in the functional tests.

Reply:
Done (`protected virtual`) and done (added unit tests of the `GCMProcess` type).